src/dataflow/src/sink/kafka.rs

```diff
@@ -85,39 +89,65 @@ pub fn kafka<G>(
             .create_with_context(SinkProducerContext)
             .expect("creating kafka producer for kafka sinks failed"),
     );
+    let mut queue: VecDeque<(Row, Diff)> = VecDeque::new();
+    let mut vector = Vec::new();
+    let mut encoded_buffer = None;
 
     let name = format!("kafka-{}", id);
     stream.sink(
         Exchange::new(move |_| sink_hash),
         &name.clone(),
         move |input| {
-            // TODO(rkhaitan): can we use a better execution model where we
-            // limit the number of inputs we consume at a time?
             input.for_each(|_, rows| {
-                for (row, _time, diff) in rows.iter() {
-                    let diff_pair = if *diff < 0 {
+                rows.swap(&mut vector);
+
+                for (row, _time, diff) in vector.drain(..) {
+                    queue.push_back((row, diff));
+                }
+            });
+
+            for _ in 0..KAFKA_SINK_FUEL {
+                let (encoded, count) = if let Some((encoded, count)) = encoded_buffer.take() {
+                    (encoded, count)
+                } else if queue.is_empty() {
+                    return;
+                } else {
+                    let (row, diff) = queue.pop_front().expect("queue known to be nonempty");
```

Comment:
Might be wrong, but can't this be `else if let Some((row, diff)) = queue.pop_front() { .. } else { return; }`?